bundles/org.eclipse.equinox.p2.director/src/org/eclipse/equinox/internal/p2/director/Slicer.java

```diff
@@ -31,42 +31,46 @@ public class Slicer {
 	private final IQueryable<IInstallableUnit> possibilites;
 	private final boolean considerMetaRequirements;
 	protected final IInstallableUnit selectionContext;
-	private final Map<String, Map<Version, IInstallableUnit>> slice; //The IUs that have been considered to be part of the problem
-	private final MultiStatus result;
+	/** The IUs that have been considered to be part of the problem */
+	private final Map<String, Map<Version, IInstallableUnit>> slice = new HashMap<>();
+	private final MultiStatus result = new MultiStatus(Slicer.class, 0, Messages.Planner_Problems_resolving_plan);
 
-	private LinkedList<IInstallableUnit> toProcess;
-	private Set<IInstallableUnit> considered; //IUs to add to the slice
-	private Set<IInstallableUnit> nonGreedyIUs = new HashSet<>(); //IUs that are brought in by non greedy dependencies
+	private Queue<IInstallableUnit> toProcess;
+	private Set<IInstallableUnit> considered; // IUs to add to the slice
+	private final Set<IInstallableUnit> nonGreedyIUs = new HashSet<>(); // IUs that are brought in by non greedy dependencies
 
 	public Slicer(IQueryable<IInstallableUnit> input, Map<String, String> context, boolean considerMetaRequirements) {
 		this(input, InstallableUnit.contextIU(context), considerMetaRequirements);
 	}
 
-	public Slicer(IQueryable<IInstallableUnit> possibilites, IInstallableUnit selectionContext, boolean considerMetaRequirements) {
+	public Slicer(IQueryable<IInstallableUnit> possibilites, IInstallableUnit selectionContext,
+			boolean considerMetaRequirements) {
 		this.possibilites = possibilites;
 		this.selectionContext = selectionContext;
 		this.considerMetaRequirements = considerMetaRequirements;
-		slice = new HashMap<>();
-		result = new MultiStatus(DirectorActivator.PI_DIRECTOR, IStatus.OK, Messages.Planner_Problems_resolving_plan, null);
 	}
 
 	public IQueryable<IInstallableUnit> slice(IInstallableUnit[] ius, IProgressMonitor monitor) {
```

Comment:
This is an internal class but I know that for example Tycho is using it.
Tycho can be adapted (and would actually benefit from less to-array conversion), but I wonder if there are any know other external user?
Should we keep this method or just delete it and only provide the replacement with an Collection argument?

Reply:
if Oomph is not using it (@merks ?) I think it can be removed.

Reply:
What is the benefit over deleting it / replacing it with a collection type method?

Reply:
Oomph doesn't use the Slicer.

Reply:
> if Oomph is not using it (@merks ?) I think it can be removed.

Removed it now.

> What is the benefit over deleting it / replacing it with a collection type method?

Less `toArray()` conversion which make the code more readable and a bit faster (but probably not measurable).

> Beside for me it has shown that sharing such code with P2 has litte to no benefit so one maybe probably better copy that over to Tycho as it is much more flexible and there are no discussions about changes here and then...

I understand and agree that discussing things in multiple repos can be annoying and slowing down development, but eventually the changes should be done up-stream (which is P2 in this case). Because otherwise one probably borrows the saved time only from the future if duplicated maintenance is necessary when new features/fixes arrive or just another developers wonders why there is a copy of an up-stream class and trys merge it again. And in worst case bug-fixes got missed.
For example this is in preparation to contribute the changes made in Tycho's https://github.com/eclipse-tycho/tycho/pull/2852 to P2.

Reply:
That a lot "maybe" and "if" that never happened in the past, instead all kind of fruitless discussions and "but what if this breaks any consumer we don't know" has arises, see also:

- https://github.com/eclipse-tycho/tycho/issues/2019

Reply:
Could be, but I don't really see it in this case.

So is everybody fine with this change or are there objections? If not, I would like to submit this.

Reply:
Seems okay.